pkg/internal/approver/allowed/evaluator.go

```diff
@@ -164,79 +168,106 @@ func (e evaluator) Subject() subjectEvaluator {
+	a       allowed
+	request *cmapi.CertificateRequest
 	sub     pkix.Name
 	allowed *policyapi.CertificateRequestPolicyAllowedX509Subject
 	fldPath *field.Path
 }
 
 func (e subjectEvaluator) Organization() field.ErrorList {
-	return evaluateSlice(e.sub.Organization, e.allowed.Organizations, e.fldPath.Child("organizations"))
+	return e.a.evaluateSlice(e.request, e.sub.Organization, e.allowed.Organizations, e.fldPath.Child("organizations"))
 }
 
 func (e subjectEvaluator) Country() field.ErrorList {
-	return evaluateSlice(e.sub.Country, e.allowed.Countries, e.fldPath.Child("countries"))
+	return e.a.evaluateSlice(e.request, e.sub.Country, e.allowed.Countries, e.fldPath.Child("countries"))
 }
 
 func (e subjectEvaluator) OrganizationalUnit() field.ErrorList {
-	return evaluateSlice(e.sub.OrganizationalUnit, e.allowed.OrganizationalUnits, e.fldPath.Child("organizationalUnits"))
+	return e.a.evaluateSlice(e.request, e.sub.OrganizationalUnit, e.allowed.OrganizationalUnits, e.fldPath.Child("organizationalUnits"))
 }
 
 func (e subjectEvaluator) Locality() field.ErrorList {
-	return evaluateSlice(e.sub.Locality, e.allowed.Localities, e.fldPath.Child("localities"))
+	return e.a.evaluateSlice(e.request, e.sub.Locality, e.allowed.Localities, e.fldPath.Child("localities"))
 }
 
 func (e subjectEvaluator) Province() field.ErrorList {
-	return evaluateSlice(e.sub.Province, e.allowed.Provinces, e.fldPath.Child("provinces"))
+	return e.a.evaluateSlice(e.request, e.sub.Province, e.allowed.Provinces, e.fldPath.Child("provinces"))
 }
 
 func (e subjectEvaluator) StreetAddress() field.ErrorList {
-	return evaluateSlice(e.sub.StreetAddress, e.allowed.StreetAddresses, e.fldPath.Child("streetAddresses"))
+	return e.a.evaluateSlice(e.request, e.sub.StreetAddress, e.allowed.StreetAddresses, e.fldPath.Child("streetAddresses"))
 }
 
 func (e subjectEvaluator) PostalCode() field.ErrorList {
-	return evaluateSlice(e.sub.PostalCode, e.allowed.PostalCodes, e.fldPath.Child("postalCodes"))
+	return e.a.evaluateSlice(e.request, e.sub.PostalCode, e.allowed.PostalCodes, e.fldPath.Child("postalCodes"))
 }
 
 func (e subjectEvaluator) SerialNumber() field.ErrorList {
-	return evaluateString(e.sub.SerialNumber, e.allowed.SerialNumber, e.fldPath.Child("serialNumber"))
+	return e.a.evaluateString(e.request, e.sub.SerialNumber, e.allowed.SerialNumber, e.fldPath.Child("serialNumber"))
 }
 
-func evaluateString(s string, crp *policyapi.CertificateRequestPolicyAllowedString, fldPath *field.Path) field.ErrorList {
-	var el field.ErrorList
-	if len(s) > 0 {
-		if crp == nil || crp.Value == nil {
-			el = append(el, field.Invalid(fldPath.Child("value"), s, "nil"))
-		} else if !util.WildcardMatches(*crp.Value, s) {
-			el = append(el, field.Invalid(fldPath.Child("value"), s, *crp.Value))
+func (a allowed) evaluateString(request *cmapi.CertificateRequest, s string, crp *policyapi.CertificateRequestPolicyAllowedString, fldPath *field.Path) field.ErrorList {
+	if len(s) == 0 {
```

Comment:
Could you add some comments in this function?
eg. explain why if len(s) == 0 we do not care about crp.Value or crp.Validations.

Reply:
Done. Might not be the best wording, but this has more to do with not breaking the policy API.